### PR TITLE
SLVS-989 SE should learn not null constraint on all cases handled in …

### PR DESCRIPTION
**Pull request by tvsonar**
…S2259

### CODE REVIEW

File: its/expected/akka.net/Akka-{5DEDDF90-37F0-48D3-A0B0-A5CBD8A7E377}-S2583.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+"issues":  [
+{
+"id":  "S2583",
+"message":  "Change this condition so that it does not always evaluate to \"false\".",
```
Comment (tvsonar) [inline its/expected/akka.net/Akka-{5DEDDF90-37F0-48D3-A0B0-A5CBD8A7E377}-S2583.json]: Nice, this issue is caused by a possible NRE.

File: src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs [hunk 2]
```diff
@@ -250,18 +248,30 @@ protected override void VisitInstruction(ExplodedGraphNode node)
...
+                            newProgramState = newProgramState
+                                .PopValue(out memberExpression)
+                                .PushValue(new MemberAccessSymbolicValue(memberExpression, memberAccess.Name.Identifier.ValueText));
                         }
                     }
                     break;
 
+                case SyntaxKind.PointerMemberAccessExpression:
+                    {
+                        var memberAccess = (MemberAccessExpressionSyntax)instruction;
```
Comment (SonarTech) [inline src/SonarLint.CSharp/Helpers/SymbolicExecution/ExplodedGraph.cs]: MAJOR 'instruction', a variable, is cast to type 'MemberAccessExpressionSyntax' multiple times in method 'ExplodedGraph.VisitInstruction(ExplodedGraphNode)'. Cache the result of the 'as' operator or direct cast in order to eliminate the redundant castclass instruction. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3ADoNotCastUnnecessarily)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: src/SonarLint.CSharp/Rules/NullPointerDereference.cs [hunk 3]
```diff
@@ -154,33 +154,66 @@ private ProgramState ProcessAwait(ProgramState programState, AwaitExpressionSynt
...
             private ProgramState ProcessMemberAccess(ProgramState programState, MemberAccessExpressionSyntax memberAccess)
             {
                 var identifier = memberAccess.Expression as IdentifierNameSyntax;
-                return ProcessAccessedIdentifier(identifier, memberAccess, programState);
+                var context = ProcessAccessedIdentifier(identifier, memberAccess, programState);
+                return SetNotNullConstraintOnSymbolIfPossible(context);
             }
 
-            private ProgramState ProcessAccessedIdentifier(IdentifierNameSyntax identifier, ExpressionSyntax expression, ProgramState programState)
+            private static ProgramState SetNotNullConstraintOnSymbolIfPossible(ProgramStateSymbolPair context)
```
Comment (benzonico) [inline src/SonarLint.CSharp/Rules/NullPointerDereference.cs]: I would remove the "IfPossible" from the name as it is not really relevant...

File: src/SonarLint.CSharp/Rules/NullPointerDereference.cs [hunk 4]
```diff
@@ -154,33 +154,66 @@ private ProgramState ProcessAwait(ProgramState programState, AwaitExpressionSynt
...
+                {
+                    return null;
+                }
+
+                if (context.LocalVarSymbol == null)
+                {
+                    return context.ProgramState;
+                }
+
+                if (!context.LocalVarSymbol.HasConstraint(ObjectConstraint.NotNull, context.ProgramState) &&
```
Comment (benzonico) [inline src/SonarLint.CSharp/Rules/NullPointerDereference.cs]: For me this should be the responsability of the SetConstraint method (returning same instance if nothing is to be done)

File: src/SonarLint.CSharp/Rules/NullPointerDereference.cs [hunk 5]
```diff
@@ -154,33 +154,66 @@ private ProgramState ProcessAwait(ProgramState programState, AwaitExpressionSynt
...
+                var context = ProcessAccessedIdentifier(identifier, elementAccess, programState);
+                return SetNotNullConstraintOnSymbolIfPossible(context);
             }
 
 
             private ProgramState ProcessMemberAccess(ProgramState programState, MemberAccessExpressionSyntax memberAccess)
             {
                 var identifier = memberAccess.Expression as IdentifierNameSyntax;
-                return ProcessAccessedIdentifier(identifier, memberAccess, programState);
+                var context = ProcessAccessedIdentifier(identifier, memberAccess, programState);
```
Comment (benzonico) [inline src/SonarLint.CSharp/Rules/NullPointerDereference.cs]: This should most probably only return the LocalVarSymbol, no need of the program state in this method.

File: src/SonarLint.CSharp/Rules/NullPointerDereference.cs [hunk 6]
```diff
@@ -189,47 +227,56 @@ private ProgramState ProcessAccessedIdentifier(IdentifierNameSyntax identifier,
...
+            }
+
+            private static ProgramState SetNotNullConstraintOnSymbol(ISymbol symbol, ProgramState programState)
+            {
+                if (programState == null)
+                {
+                    return null;
+                }
+
+                if (symbol == null)
```
Comment (benzonico) [inline src/SonarLint.CSharp/Rules/NullPointerDereference.cs]: matter of style here but I would prefer the more concise : 

```
if(programState == null || symbol == null) {
  return programState;
}
```

**Pull request merged**